fix `sampleTerrain` from exploding if RequestScheduler throttles it 💣

DanielLeone opened the pull request:
Bug: If you call `sampleTerrain` with positions spread over too large an area, too many tile requests may be made, and if `throttleByServer` is set to `true` for the request, eventually `requestTileGeometry` will return `undefined` and explode the whole task (unhandled exception).

### The error I'm seeing
![image](https://user-images.githubusercontent.com/8600684/171794245-22bc5893-2dc1-42a3-aae8-00350f304bb1.png)

[The sandcastle to reproduce](https://sandcastle.cesium.com/#c=jVdbb9s2FP4rnJ9k1KEs3+WkxZI0a4O1S1EbK4Z6D7R0LBGlSYGknLhD/vuOrpZsp+1LQpHn8p37caCksYTp4B03n7Ta8RA0eU0kPJJbMDzd0mt8u18suYDwTsCOWa7kErRmXFYMzn8rSUiqxTz7T8iqE1ubmLnrQsUxDCkqibihgdq6xdHVYKxrQO94AMb9orQ4qBi+dUsteLrfsggWSAh61emt5HP3ciVXMsjB7zg8HoP+O79zVp0g/75V0qKknJvkYG3bgvmRCw46VlKArcgXbJsI+KQMzyCanGCTyiD7IkKpb9f2o71DlGiY0y00FSAt0xHKaYG8ZdriicmhU/ht2O9Pp3067s+mg+HMH85GveJhPBmOfG9AvaE39Efj2bi8H/jTmTce0MFsMh6Pxp43ye4z4JVetdmYn+pF6SPqT0d9fzaZzsaTQSn+YuoP6XDk+VN/1p9505FX6R35PvUR52zSn06G3mRw0FuEgwZsC5rRwilOYX6vhPMy4VIzaTZKb50S60dmNX8a0fu3d38t75f/ZKzPLbebPCZlrizSACNunPPxakUEhOCJUTxE55TK7qor+uXdYjbKw0uI697sSQgblgq0oCAloQJDpEIHr02qgUSgtmD1vmBYQ8xlWGUNJQuwlsuI2JgbshEsIiDZWqAIGzNLG94wAUigkVBroCEkNl5iKl1HKMXY0kbEa3UKJbqSD6RFG8FQk5oEZIhJKK1xWp6uaTRs1Q6uhXC6pRR0OXGyROcovX+J/65eyHkqQEY2viSvXvHSn5VHk5ImA3iW9yv/t9R3iomFoVNKI0RiTsxreTQGHsWWWvUHf4LQ8bq9irAimR/CiQmlrYo0S2IeLFWd7E5Fe+BecyHWiulwTmrViAy05QETD5pHHCVXLaV1TW8elsuHj70Dm8E3BN2n08YlzxrXHBsipW5+Nu6GBVxwu6cR3+TNrKB8rk+CrUG0EFl4sr/kjiyQ0mb6vH5iyVZJZRIWQEMPIbHS/HvWEI8tfH/0QG+x5u4+N1gTVCge8hqen+0oA6ffIxfeqAUJvXyrhNK1ovyL3ny4vv2zQadSK7BH/wqpidXjDQu+RVqlEqOX1UPjeV0/lcKaULMbp0/9Hmn+mXbP8n/CtMTCfcnYUY8MuychfC6q7vl86Zl0C43qbDezQAOz8E7z0NEQWCYjAe+Z2Cz4d2h1rwhJvvDQxlhsI++y/fA+T5DjFyjm0gccrzYN4dD2sMfGmE2fWYhGGWcwpf5sNhhPu2eYlYx+zD2bUH8w9kbjFnd7AH2uTCvnz4nwC3Jifa9NWhnxc8pa6KtfFnqG8niunm9waObXfy9bDXVfNNQ9NtRDaLLuua+7Z037VNA+lbR5fDPSp+6hGxTqxQtxEKAT55DIQB/Rrl7zAljr4om4xDkk0wXxutVjYW5D5/nEOVFpsJDjlk6pdOtmXyktE/VHWhtD5agGqx7v1M7o1RgPUl6YYklq4sNEKKmf67LVYFMtX1778qpdMBkG6E4B2fRaKiXWTN+k1iqZO2TVKRjJR4WGvAVcQXGJJuV+SMpxXvbmugU4dbD5hji/lQbj/oG9MXTKfnK0vlK2Q9lsnQ+WbiNbHnENUY8UJ5O2jSCtOq2tKcNXw+PFXmPSJMGw4UWWnjYGYnCZD7KLUjlJSu2rzmnwCgc2/frTymn0vj7t98eVMDRyyw1krVOJHdRmlJ550ZKGvee91mtO2HOgarOKQ5eiG6RzbuGssJ78BigmQa9Ih4KzyBCTea1Yv84xrWSn17kydi/gTQXjd77NQpL90HJwn7CA0tBjxl2nwTewNDjguHKbrFch3xEevj7zc4gEghmDL5tUiLzRdd5cuUh/wioUy0bhA4IUbJ+Rxd6bD8UlpfTKxc/znLaoDGSpLDm8HbmkLaFx/h8) (just hit the Sample Terrain button)


### The lines of code causing the actual issue:
`requestTileGeometry` will return `undefined` if the throttling takes place by `RequestScheduler`
https://github.com/CesiumGS/cesium/blob/f40912628795422f42fbc4db586daa38bf6e7f87/Source/Core/sampleTerrain.js#L85-L92


### The spot that `throttleByServer` is enforced to `true`, for ArcGIS terrain at least
https://github.com/CesiumGS/cesium/blob/f40912628795422f42fbc4db586daa38bf6e7f87/Source/Core/ArcGISTiledElevationTerrainProvider.js#L641-L645


### The bit of `RequestScheduler` which is doing the throttling 
https://github.com/CesiumGS/cesium/blob/f40912628795422f42fbc4db586daa38bf6e7f87/Source/Core/RequestScheduler.js#L402-L409



#### Left to do
If you guys are happy with the implementation, I'll button it up with some tests, a mention in the changelog, remove the debug logs, and revert the sandcastle change. I haven't yet as the implementation is questionable.

Cheers! 👍

cesium-concierge commented:
Thank you so much for the pull request @DanielLeone! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/DanielLeone/cesium/blob/fix-sample-terrain-throttle-error/CONTRIBUTORS.md) file!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/fix-sample-terrain-throttle-error/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/main/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

ggetz commented on Source/Core/sampleTerrain.js:
```diff
@@ -48,6 +48,78 @@ function sampleTerrain(terrainProvider, level, positions) {
   });
 }
 
+const ACTION__CONTINUE = "continue";
```
CesiumJS uses enums for strings like this. However, since there's only two possible states, maybe we could resolve to true or false depending on the state?

ggetz commented on Source/Core/sampleTerrain.js:
```diff
@@ -48,6 +48,78 @@ function sampleTerrain(terrainProvider, level, positions) {
+  if (!requestPromise) {
+    // Uh, we probably got throttled client side?
+    //  let's wait 100ms, then try again
+    // This will turn into an infinite loop if we don't eventually
```
This is true of other areas of the code which use the request scheduler. I wouldn't be too concerned, as the only situation which this returns undefined should be one where a retry is needed.

ggetz commented on Source/Core/sampleTerrain.js:
```diff
@@ -78,23 +150,14 @@ function doSampling(terrainProvider, level, positions) {
+  return drainTileRequestQueue(tileRequests, tilePromises).then(function () {
+    log("all requests made, waiting for all responses");
+    return Promise.all(tilePromises).then(function () {
+      log("all inflight requests are complete, sampling done");
```
Can we remove the logging in this class? It's useful for debugging, but won't be useful to the end user.

ggetz commented on Source/Core/sampleTerrain.js:
```diff
@@ -48,6 +48,78 @@ function sampleTerrain(terrainProvider, level, positions) {
+  );
+  // nothing to do, return
+  if (!tileRequests.length) {
+    return Promise.resolve(null);
```
We use `undefined` instead of `null`, so this can be simplified to `Promise.resolve()`.

ggetz commented on Source/Core/sampleTerrain.js:
```diff
@@ -48,6 +48,78 @@ function sampleTerrain(terrainProvider, level, positions) {
+
+function delay(ms) {
+  return new Promise(function (res) {
+    window.setTimeout(res, ms);
```
```suggestion
    setTimeout(res, ms);
```

ggetz commented:
Thanks @DanielLeone! Looks like we have your CLA, but could you add your username to the [contributors](https://github.com/DanielLeone/cesium/blob/fix-sample-terrain-throttle-error/CONTRIBUTORS.md) file!

DanielLeone commented:
@ggetz Thanks for the quick review and good feedback!
* I've actioned all the feedback items
* added a unit test
* added a line to CHANGES.md (and fixed an old URL typo in there my IDE was complaining about)

My name is already in contributors file, not sure why the bot isn't picking that up
https://github.com/CesiumGS/cesium/blob/74dac36b6b522314d0ec7eccf7c934ebf12cb797/CONTRIBUTORS.md?plain=1#L285

ggetz reviewed (commented):
Awesome, thanks @DanielLeone! I just have one more comment. Thanks for adding the unit test and responding to feedback. :+1:

DanielLeone commented on Source/Core/sampleTerrain.js:
```diff
@@ -48,6 +48,79 @@ function sampleTerrain(terrainProvider, level, positions) {
 }
 
+/**
+ * @param {Array.<*>} tileRequests The mutated list of requests, the first one will be attempted
+ * @param {Array.<Promise<*>>} results The list to put the result promises into
```
correct, `Promise<Object>` sounds good to me; and I've changed the 2nd line to be `Promise<void>` which I believe always the case in this code path. 👍 Changes made

ggetz commented:
Thanks @DanielLeone!
